src/frontend/src/optimizer/rule/distinct_agg.rs

```diff
@@ -34,29 +36,16 @@ impl Rule for DistinctAggRule {
         let agg: &LogicalAgg = plan.as_logical_agg()?;
         let (mut agg_calls, mut agg_group_keys, input) = agg.clone().decompose();
         let original_group_keys_len = agg_group_keys.len();
-        let (distinct_aggs, non_distinct_aggs): (Vec<_>, Vec<_>) = agg_calls
-            .clone()
-            .into_iter()
-            .partition(|agg_call| agg_call.distinct);
-        if distinct_aggs.is_empty() {
-            return None;
-        }
-        let flag_value_of_distinct_agg = if non_distinct_aggs.is_empty() { 0 } else { 1 };
-        let input_schema_len = input.schema().len();
 
-        let expand = Self::build_expand(input, &agg_group_keys, &distinct_aggs, &non_distinct_aggs);
-        let project = Self::build_project(
-            input_schema_len,
-            expand,
-            &mut agg_group_keys,
-            &mut agg_calls,
-        );
-        let mid_agg = Self::build_middle_agg(project, agg_group_keys, agg_calls.clone());
+        let (node, flag_values, has_expand) =
+            Self::build_expand(input, &mut agg_group_keys, &mut agg_calls)?;
+        let mid_agg = Self::build_middle_agg(node, agg_group_keys, agg_calls.clone(), has_expand);
         Some(Self::build_final_agg(
             mid_agg,
             original_group_keys_len,
             agg_calls,
-            flag_value_of_distinct_agg,
+            flag_values.into_iter(),
```

Comment:
We can pass `Vec` directly and create the iterator inside.

When used as function parameter, `std::vec::IntoIter` should be replaced with `impl Iterator<Item = usize>`. But it is unnecessary here.